<a href="https://colab.research.google.com/github/ulischlickewei/DS-AI/blob/main/Week1/20240408_Short_introduction_to_Jupyter_Notebooks_and_Python_solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Short introduction to Jupyter Notebooks and to Python

In today's lab, you'll learn/review how to:

* Navigate Jupyter notebooks (like this one)
* Write and evaluate some basic expressions in Python, the computer language of the course and use Python as a calculator
* Create a variable in Python
* Import code other people have written
* Use conditional statements in Python
* Do iterations in Python
* Write function in Python


*Copyright note*: The material in this lab is taken in significant parts from [Data 8 class of Berkeley University](http://data8.org/). These materials are generously shared by the colleagues under a Creative Commons license.

## 0) Jupyter Notebooks and Google Colab

### 0.1: Jupyter Notebooks

A Jupyter Notebook is a basically a nice looking web interface for the interactive Python interpreter. Interactive coding differs from classical coding. Instead of executing the whole program at once, you can execute small code cells in a sequential way. You can think of Jupyter Notebook as a chalk board. With each code cell, we we write something onto the chalk board. As long as we don't restart Notebook (=wiping the chalk board), we keep the changes made to our Notebook. This especially means that we don't have to execute previous code cells again and again if we make changes in subsequent cells. This also means that we have to execute previous code cells again (=overwriting the text written onto the chalk board) if we made changes to them.

Jupyter Notebooks also allow text cells. With the text cells, we can create interactive code stories in which you can play around and experiment with the theory provided in the text cell. There even exist whole books written as Jupyter Notebooks!

#### Text cells
In a notebook, each rectangle containing text or code is called a *cell*.

Text cells (like this one) can be edited by double-clicking on them. They're written in a simple format called [Markdown](http://daringfireball.net/projects/markdown/syntax) to add formatting and section headings.  You don't need to learn Markdown, but you might want to.

After you edit a text cell, click the "run cell" button at the top that looks like ▶| or hold down `shift` + `return` to confirm any changes. (Try not to delete the instructions of the lab.)

#### Code cells
Other cells contain code in the Python 3 language. Running a code cell will execute all of the code it contains.

To run the code in a code cell, first click on that cell to activate it.  It'll be highlighted with a little green or blue rectangle.  Next, either press ▶| or hold down `shift` + `return`.

Try running this cell:

In [ ]:
print("Hello world!")

Hello world!


## 0.2: Google Colab

Google Colab is a freemium cloud service based on Jupyter Notebooks. It provides a virtual machine (VM) in which a Jupyter Notebook can be executed, even with hardware acceleration like GPUs. The whole process is similar to Google borrowing us one of its computers. We can access this borrowed computer through this browser window which shows a Jupyter Notebook running on Google's servers.

## 1) Python basics

Python is a high-level programming language with a very user-friendly syntax. Many open-source tools for Scientific Computing, Data Science and Machine Learning are available in Python which makes it one of the most important programming languages for scientits and engineers.

Let's quickly browse through some of the basics of Python.

### 1.1: Expressions
The fundamental building block of Python code is an expression. Cells can contain multiple lines with multiple expressions. When you run a cell, the lines of code are executed in the order in which they appear. Every `print` expression prints a line. Run the next cell and notice the order of the output.

In [ ]:
print("First this line is printed,")
print("and then this one.")

First this line is printed,
and then this one.


### 1.2: Using Python as a calculator

Quantitative information arises everywhere in Mathematics. In addition to representing commands to print out lines, expressions can represent numbers and methods of combining numbers. The expression `3.2500` evaluates to the number 3.25. (Run the cell and see.)

In [ ]:
3.2500

3.25

The line in the next cell subtracts.  Its value is what you'd expect.  Run it.

In [ ]:
3.25 - 1.5

1.75

Many basic arithmetic operations are built into Python.  The Data 8 textbook section on [Expressions](http://www.inferentialthinking.com/chapters/03/1/Expressions.html) describes the most important arithmetic operators.  The common operator that differs from typical math notation is `**`, which raises one number to the power of the other. So, `2**3` stands for $2^3$ and evaluates to 8.

The order of operations is the same as what you learned in elementary school, and Python also has parentheses.  For example, compare the outputs of the cells below. The second cell uses parentheses for a happy new year!

In [ ]:
5+6*5-6*3**2*2**3/4*7

-721.0

In [ ]:
5+(6*5-(6*3))**2*((2**3)/4*7)

2021.0

In standard math notation, the first expression is

$$3 + 6 \times 5 - 6 \times 3^2 \times \frac{2^3}{4} \times 7,$$

while the second expression is

$$3 + (6 \times 5 - (6 \times 3))^2 \times (\frac{(2^3)}{4} \times 7).$$

##### Question 1.2.1.
Write a Python expression in this next cell that's equal to $5 \times (3 \frac{10}{11}) - 50 \frac{1}{3} + 2^{.5 \times 22} - \frac{7}{33} + 4$.  That's five times three and ten elevenths, minus fifty and a third, plus two to the power of half twenty-two, minus seven thirty-thirds plus four.  By "$3 \frac{10}{11}$" we mean $3+\frac{10}{11}$, not $3 \times \frac{10}{11}$.

Replace the ellipses (`...`) with your expression.  Try to use parentheses only when necessary.

*Hint:* The correct output should start with a familiar number.

In [ ]:
### SOLUTION
5*(3+10/11) - (50+1/3) + 2**(.5*22) - 7/33 + 4

2021.0

### 1.3 Names/variables

In natural language, we have terminology that lets us quickly reference very complicated concepts.  We don't say, "That's a large mammal with brown fur and sharp teeth!"  Instead, we just say, "Bear!"

In Python, we do this with *assignment statements*. An assignment statement has a name on the left side of an `=` sign and an expression to be evaluated on the right.

In [ ]:
ten = 3 * 2 + 4

When you run that cell, Python first computes the value of the expression on the right-hand side, `3 * 2 + 4`, which is the number 10.  Then it assigns that value to the name `ten`.  At that point, the code in the cell is done running.

After you run that cell, the value 10 is bound to the name `ten`:

In [ ]:
ten

10

A common pattern in Jupyter notebooks is to assign a value to a name and then immediately evaluate the name in the last line in the cell so that the value is displayed as output.

In [ ]:
close_to_pi = 355/113
close_to_pi

3.1415929203539825

##### Question 1.3.1.
Newton's law of universal gravitation predicts that any object near the surface of the Moon which falls freely travels a distance of

$$s = \frac{1}{2} G \frac{M}{R^2} t^2 \text{ meters}$$

where $t$ is the time of fall in seconds, $G$ is a universal constant, $M$ is the moon's mass in kilograms, and $R$ is the moon's radius in meters.  So if we know $G$, $M$, and $R$, then Newton's laws let us predict how far an object will fall over any amount of time.

A hammer falls from a height of 113cm to the surface on the Moon. The falls takes 1.2 seconds.

Complete the following code cell to compute the difference between predicted and measured distance that the hammer fell in this experiment.

In [ ]:
# t, the duration of the fall in the experiment, in seconds.
# **YOU FILL THIS VALUE IN**
### SOLUTION
time = 1.2

# The estimated distance the hammer actually fell, in meters.
# **YOU FILL THIS VALUE IN**
### SOLUTION
estimated_distance_m = 1.13

# Next, we've written down the values of the 3 universal constants
# that show up in Newton's formula.

# G, the universal constant measuring the strength of gravity.
gravity_constant = 6.674 * 10**-11

# M, the moon's mass, in kilograms.
moon_mass_kg = 7.34767309 * 10**22

# R, the radius of the moon, in meters.
moon_radius_m = 1.737 * 10**6

# The distance the hammer should have fallen
# over the duration of the fall, in meters,
# according to Newton's law of gravity.
# The text above describes the formula
# for this distance given by Newton's law.
# **YOU FILL THIS PART IN.**
### SOLUTION
predicted_distance_m = 1/2 * gravity_constant * moon_mass_kg / moon_radius_m**2 * time**2

# Here we've computed the difference
# between the predicted fall distance and the distance we actually measured.
# If you've filled in the above code, this should just work.
difference = predicted_distance_m - estimated_distance_m
difference

0.040223694659304865

### 1.4: Importing code

Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Rather than copy-pasting, Python allows us to **import modules**. A module is a file with Python code that has defined variables and functions. By importing a module, we are able to use its code in our own notebook.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python.

Suppose we want to very accurately compute the area of a circle with a radius of 5 meters.  For that, we need the constant $\pi$, which is roughly 3.14.  Conveniently, the `math` module has `pi` defined for us:

In [ ]:
import math
radius = 5
area_of_circle = radius**2 * math.pi
area_of_circle

78.53981633974483

In the code above, the line `import math` imports the math module. This statement creates a module and then assigns the name `math` to that module. We are now able to access any variables or functions defined within `math` by typing the name of the module followed by a dot, then followed by the name of the variable or function we want.

    <module name>.<name>

##### Question 1.4.1.
The module `math` also provides the name `e` for the base of the natural logarithm, which is roughly 2.71. Compute $e^{\pi}-\pi$, giving it the name `near_twenty`.

*Remember: You can access `pi` from the `math` module as well!*

In [ ]:
### SOLUTION
near_twenty = math.e**(math.pi) - math.pi
near_twenty

19.99909997918947

![XKCD](http://imgs.xkcd.com/comics/e_to_the_pi_minus_pi.png)

[Source](http://imgs.xkcd.com/comics/e_to_the_pi_minus_pi.png)
[Explanation](https://www.explainxkcd.com/wiki/index.php/217:_e_to_the_pi_Minus_pi)

## 2) Functions

### 2.1: Calling functions

The most common way to combine or manipulate values in Python is by calling functions. Python comes with many built-in functions that perform common operations.

For example, the `abs` function takes a single number as its argument and returns the absolute value of that number. Run the next two cells and see if you understand the output.

In [ ]:
abs(5)

5

In [ ]:
abs(-5)

5

In Question 1.4.1 above, you accessed variables within the `math` module.

**Modules** also define **functions**.  For example, `math` provides the name `floor` for the floor function.  Having imported `math` already, we can write `math.floor(7.5)` to compute the floor of 7.5.  (Note that the floor function returns the largest integer less than or equal to a given number.)

##### Question 2.1.1.
Compute the floor of pi using `floor` and `pi` from the `math` module.  Give the result the name `floor_of_pi`.

In [ ]:
floor_of_pi = math.floor(math.pi)
floor_of_pi

3

###2.2: Defining functions

Let's write a very simple function that converts a proportion to a percentage by multiplying it by 100.  For example, the value of `to_percentage(.5)` should be the number 50 (no percent sign).

A function definition has a few parts.

##### `def`
It always starts with `def` (short for **def**ine):

    def

##### Name
Next comes the name of the function.  Like other names we've defined, it can't start with a number or contain spaces. Let's call our function `to_percentage`:
    
    def to_percentage

##### Signature
Next comes something called the *signature* of the function.  This tells Python how many arguments your function should have, and what names you'll use to refer to those arguments in the function's code.  A function can have any number of arguments (including 0!).

`to_percentage` should take one argument, and we'll call that argument `proportion` since it should be a proportion.

    def to_percentage(proportion)
    
If we want our function to take more than one argument, we add a comma between each argument name. Note that if we had zero arguments, we'd still place the parentheses () after that name.

We put a **colon** after the signature to tell Python that the next indented lines are the body of the function. If you're getting a syntax error after defining a function, check to make sure you remembered the colon!

    def to_percentage(proportion):

##### Documentation
Functions can do complicated things, so you should write an explanation of what your function does.  For small functions, this is less important, but it's a good habit to learn from the start.  Conventionally, Python functions are documented by writing an **indented** triple-quoted string:

    def to_percentage(proportion):
        """Converts a proportion to a percentage."""

##### Body
Now we start writing code that runs when the function is called.  This is called the *body* of the function and every line **must be indented with a tab**.  Any lines that are *not* indented and left-aligned with the def statement is considered outside the function.

Some notes about the body of the function:
- We can write code that we would write anywhere else.  
- We use the arguments defined in the function signature. We can do this because we assume that when we call the function, values are already assigned to those arguments.
- We generally avoid referencing variables defined *outside* the function. If you would like to reference variables outside of the function, pass them through as arguments!


Now, let's give a name to the number we multiply a proportion by to get a percentage:

    def to_percentage(proportion):
        """Converts a proportion to a percentage."""
        factor = 100

##### `return`
The special instruction `return` is part of the function's body and tells Python to make the value of the function call equal to whatever comes right after `return`.  We want the value of `to_percentage(.5)` to be the proportion .5 times the factor 100, so we write:

    def to_percentage(proportion):
        """Converts a proportion to a percentage."""
        factor = 100
        return proportion * factor
        
`return` only makes sense in the context of a function, and **can never be used outside of a function**. `return` is always the last line of the function because Python stops executing the body of a function once it hits a `return` statement. If a function does not have a return statement, it will not return anything; if you expect a value back from the function, make sure to include a return statement.

*Note:*  `return` inside a function tells Python what value the function evaluates to. However, there are other functions, like `print`, that have no `return` value. For example, `print` simply prints a certain value out to the console.

`return` and `print` are **very** different.

##### Question 2.2.1.
Define `to_percentage` in the cell below.  Call your function to convert the proportion .2 to a percentage.  Name that percentage `twenty_percent`.

In [ ]:
### SOLUTION
def to_percentage(proportion):
  factor = 100
  return proportion * factor

twenty_percent = to_percentage(0.2)
twenty_percent

20.0

Here's something important about functions: the names assigned *within* a function body are only accessible within the function body. Once the function has returned, those names are gone.  So even if you created a variable called `factor` and defined `factor = 100` inside of the body of the `to_percentage` function and then called `to_percentage`, `factor` would not have a value assigned to it outside of the body of `to_percentage`:

In [ ]:
factor

NameError: ignored

Like you’ve done with built-in functions in previous labs (max, abs, etc.), you can pass in named values as arguments to your function.

##### Question 2.2.2.
Use `to_percentage` again to convert the proportion named `a_proportion` (defined below) to a percentage called `a_percentage`.

*Note:* You don't need to define `to_percentage` again!  Like other named values, functions stick around after you define them.

In [ ]:
a_proportion = 2**(0.5)/2
a_percentage = to_percentage(a_proportion)
a_percentage

70.71067811865476

As we've seen with built-in functions, functions can also take strings (or arrays, or tables) as arguments, and they can return those things, too.

In the following cell, we will define a function called `disemvowel`.  It takes in a single string as its argument. It returns a copy of that string, but with all the characters that are vowels removed.  (In English, the vowels are the characters "a", "e", "i", "o", and "u".)

To remove all the "a"s from a string, we used `a_string.replace("a", "")`.  The `.replace` method for strings returns a new string, so we can call `replace` multiple times, one after the other.

In [ ]:
def disemvowel(a_string):
    """Removes all vowels from a string."""
    return a_string.replace("a", "").replace("e", "").replace("i", "").replace("o", "").replace("u", "")

# An example call to the function.  (It's often helpful to run
# an example call from time to time while we're writing a function,
# to see how it currently works.)
disemvowel("Can you read this without vowels?")

'Cn y rd ths wtht vwls?'

##### Calls on calls on calls
Just as you write a series of lines to build up a complex computation, it's useful to define a series of small functions that build on each other.  Since you can write any code inside a function's body, you can call other functions you've written.

If a function is a like a recipe, defining a function in terms of other functions is like having a recipe for cake telling you to follow another recipe to make the frosting, and another to make the jam filling.  This makes the cake recipe shorter and clearer, and it avoids having a bunch of duplicated frosting recipes.  It's a foundation of productive programming.

For example, suppose you want to count the number of characters *that aren't vowels* in a piece of text.  One way to do that is this to remove all the vowels and count the size of the remaining string.

##### Question 2.2.3.
Write a function called `num_non_vowels`.  It should take a string as its argument and return a number.  That number should be the number of characters in the argument string that aren't vowels. You should use the `disemvowel` function we provided above inside of the `num_non_vowels` function.

*Hint:* The function `len` takes a string as its argument and returns the number of characters in it.

In [ ]:
### SOLUTION
def num_non_vowels(a_string):
    """The number of characters in a string, minus the vowels."""
    n = len(disemvowel(a_string))
    return n

# Try calling your function yourself to make sure the output is what
# you expect.

s = "argument"
num_non_vowels(s)

5

## 3) Conditional statements

A conditional statement is a multi-line statement that allows Python to choose among different alternatives based on the truth value of an expression.

Here is a basic example.

```
def sign(x):
    if x > 0:
        return 'Positive'
    else:
        return 'Negative'
```

If the input `x` is greater than `0`, we return the string `'Positive'`. Otherwise, we return `'Negative'`.

If we want to test multiple conditions at once, we use the following general format.

```
if <if expression>:
    <if body>
elif <elif expression 0>:
    <elif body 0>
elif <elif expression 1>:
    <elif body 1>
...
else:
    <else body>
```

Only the body for the first conditional expression that is true will be evaluated. Each `if` and `elif` expression is evaluated and considered in order, starting at the top. `elif` can only be used if an `if` clause precedes it. As soon as a true value is found, the corresponding body is executed, and the rest of the conditional statement is skipped. If none of the `if` or `elif` expressions are true, then the `else body` is executed.

For more examples and explanation, refer to the section on conditional statements [here](https://inferentialthinking.com/chapters/09/1/Conditional_Statements.html).

##### Question 3.1.
Implement a function `my_abs` which takes a number as an argument and returns the absolute value of that number.

In [ ]:
### SOLUTION
def my_abs(x):
  if x >= 0:
    return x
  else:
    return -x

print('Absolute value of 5 = ', my_abs(5))
print('Absolute value of -5 = ', my_abs(-5))

Absolute value of 5 =  5
Absolute value of -5 =  5


## 4) Iterations

Using a `for` statement, we can perform a task multiple times. This is known as iteration. The general structure of a for loop is:

`for <placeholder> in <array>:` followed by indented lines of code that are repeated for each element of the `array` being iterated over. You can read more about for loops [here](https://www.inferentialthinking.com/chapters/09/2/Iteration.html).

**NOTE:** We often use `i` as the `placeholder` in our class examples, but you could name it anything! Some examples can be found below.

One use of iteration is to loop through a set of values. For instance, we can print out all of the colors of the rainbow.

In [ ]:
rainbow = ["red", "orange", "yellow", "green", "blue", "indigo", "violet"]

for color in rainbow:
    print(color)

red
orange
yellow
green
blue
indigo
violet


We can see that the indented part of the `for` loop, known as the body, is executed once for each item in `rainbow`. The name `color` is assigned to the next value in `rainbow` at the start of each iteration. Note that the name `color` is arbitrary; we could easily have named it something else. The important thing is we stay consistent throughout the `for` loop.

In [ ]:
for another_name in rainbow:
    print(another_name)

red
orange
yellow
green
blue
indigo
violet


Typically, we will iterate over integer numbers. For example:

In [ ]:
for i in range(10):
  print(i)

0
1
2
3
4
5
6
7
8
9


Here, we are using the [range](https://docs.python.org/3/library/stdtypes.html#typesseq-range) function from Python.


##### Question 4.1.
Write a function `my_factorial(n)` that computes $n! = n \cdot (n-1) \cdot \ldots \cdot 1$ for a given natural number $n$.

*Hint*: You need to make sure that your `range()` in the iteration starts at the right number. See the [documentation](https://docs.python.org/3/library/stdtypes.html#typesseq-range).

In [ ]:
### SOLUTION
def my_factorial(n):
  result = 1
  for i in range(n):
    result = result*(i+1)
  return result

my_factorial(5)

120

Obviously, Python provides a module which implements the factorial function, the `math` module:

In [ ]:
import math
math.factorial(5)

120